### PR TITLE
Add two functions and spago support

### PR DESCRIPTION
**Pull request by flip111**
* Levenshtein distance
* Sørensen–Dice coefficient

Don't know how documentation update works.

### CODE REVIEW

**Comment by thomashoneyman** [comment]
Hi @flip111! Thanks for adding these. I'm happy to add in the implementations, but I'm not quite ready to move the project over to Spago (we're going to need to move all the contributor libraries at some point, and I'd like to do them at once so that we can set up automatic publishing of new releases and that sort of thing).

To upload new documentation, see the [publishing section](https://github.com/purescript-contrib/pulp#releasing-packages) in the Pulp docs.

With that in mind, could you revert the Spago parts of this change and remain on Bower for the dependencies?

File: src/Data/String/Extra.purs
Comment (thomashoneyman) [inline src/Data/String/Extra.purs]: Can we implement these using snake case?

```
sorensenDiceCoefficient
```

Also, the example still uses "levenshtein" here.

File: src/Data/String/Extra.js
Comment (thomashoneyman) [inline src/Data/String/Extra.js:1]: I'm OK with these functions being implemented in the FFI for performance reasons. However, if you believe it's possible to convert the implementation to PureScript without severe performance penalties that's always the preferred default. Given the choice between not having these at all and having the FFI implementations we can use the FFI here.

File: src/Data/String/Extra.purs
Comment (flip111) [reply src/Data/String/Extra.purs]: Yes ok we can use `camelCase` and not `snake_case`, it stays in the desert anyway. The example function name is wrong but the result is correct. Can you make direct edits instead of comments so i can approve them?

**Comment by flip111** [comment]
> we're going to need to move all the contributor libraries at some point, and I'd like to do them at once so that we can set up automatic publishing of new releases and that sort of thing

I don't know what this means. But i get that you want to do stuff before adding this to spago.

I added spago because i use these two functions myself and i use spago for my project so i like this package to be on spago. What would be the difficulty adding it to spago now and then later do these other things?

File: src/Data/String/Extra.js
Comment (flip111) [reply src/Data/String/Extra.js:1]: I don't believe that's going to benefit performance. The API to purescript is pretty clean at the moment with a pure function interface.

**Comment by flip111** [comment]
So you want `packages.dhall` and `spago.dhall` removed ?

**Comment by thomashoneyman** [comment]
> I added spago because i use these two functions myself and i use spago for my project so i like this package to be on spago. What would be the difficulty adding it to spago now and then later do these other things?

This package is already in the official package sets (see: https://github.com/purescript/package-sets/blob/e689c7da77b3b0589eb8afd52d905099dd09dfad/src/groups/purescript-contrib.dhall#L208), so you can install it with `spago install strings-extra`.

I just don't want to switch the code here to use Spago as the dependency manager because we are going to switch purescript-contrib projects to use Spago as their dependency manager in the near-ish future, and I'd like to switch them all at the same time. We're going to have to switch things like our publishing workflow when we do that, and it's just easier to do it all at once (as a maintainer).

But that doesn't affect anyone installing this library into their project -- they can still do that with Spago no problem. It's purely an internal thing.

> So you want packages.dhall and spago.dhall removed ?

Yes, thank you!

**Comment by thomashoneyman** [comment]
Looks like the CI failure is from a missing dependency (in Bower):

```
$ npm run -s test

* Building project in /home/travis/build/purescript-contrib/purescript-strings-extra

[1/1 ModuleNotFound] test/Main.purs:8:1

  8  import Data.Number.Approximate ((≅))
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  Module Data.Number.Approximate was not found.

  Make sure the source file exists, and that it has been provided as an input to the compiler.
```

I see that this module is from `sharkdp/purescript-numbers`; I'm sorry to have another request, but as a general rule the `purescript-contrib` libraries only have dependencies on either `purescript` libraries or other `purescript-contrib` libraries.

That's because when there's a breaking change to the ecosystem (for example, the new 0.14 compiler release) we need to make sure we can apply the updates to all PureScript and PureScript Contrib libraries without being blocked by maintainers that might be out of town, on vacation, unresponsive, etc.

Are you able to either a) inline this function into the test source code or b) test this equality some other way, like a regular `==` or by multiplying to make the results integers and compare those instead? Just something that won't incur the dependency on a non-contrib library.

File: test/Main.purs
Comment (thomashoneyman) [inline test/Main.purs]: To remove the dependency on `Data.Number.Approximate`, do you think we could do something like this?

```purs
normalize :: Number -> Int
normalize n = Int.round (n * 100000.0)

normalizedCoefficient :: String -> String -> Int
normalizedCoefficient a b = normalize (String.sorensenDiceCoefficient a b)

log "sorensenDiceCoefficient" *> do
  assert $ (normalizedCoefficient "CONSERVATIONALISTS" "CONVERSATIONALISTS") == (normalize 0.7647059)
```

**Review by thomashoneyman — APPROVED** [review]

**Pull request merged**
